### ISSUE TITLE
Finetuning with limited batchsize

### ISSUE THREAD

**Issue by PixarHero** [TIGER-AI-Lab/VLM2Vec#112]
I'm using a small batch size (16) due to limited computational resources, and I noticed the loss becomes zero. I’m wondering — is there a minimum batch size threshold below which training fails to work effectively? Thanks!

**Comment by memray** [comment]
Hi @PixarHero ,

I think the minimum batch size varies by dataset. Two methods might mitigate this:
(1) Simply using a larger batch size. If OOM is the concern, you can consider setting a smaller `--gc_q_chunk_size/gc_k_chunk_size` (gradcache size) or lower resolution `--resize_max_pixels`.
(2) Set a larger --interleave_batch_size (up to batch size) to make the whole batch more homogeneous (sampled from the same dataset). 
Hard negative mining is another fundamental way to mitigate this problem and we may add this feature in the near future.

Thanks,
Rui

**Comment by PixarHero** [comment]
`    "configurations": [
        {
            "name": "Python: TorchRun",
            "type": "debugpy",
            "request": "launch",
            // "program": "${file}",  # /home/ph/anaconda3/envs/alpaca-lora/lib/python3.8/site-packages/torch/distributed/run.py
            "console": "integratedTerminal",
            "module": "torch.distributed.run",
            "justMyCode": true,
            "env": {
                "MASTER_ADDR": "localhost",
                "MASTER_PORT":"12355",
                "CUDA_VISIBLE_DEVICES": "0,1"
            },
            "args": [
                "--nnodes", "1",
                "--nproc_per_node", "2",
                "--node_rank", "0",
                "--master-addr", "localhost",
                "--master-port", "12355",
                "train.py",
                "--model_name", "Qwen/Qwen2-VL-2B-Instruct",
                "--model_backbone", "qwen2_vl",
                "--dataset_name", "TIGER-Lab/MMEB-train",
                "--subset_name", "MSCOCO_t2i", "MSCOCO_i2t",
                "--num_sample_per_subset", "10000",
                "--image_dir", “vlm2vec_train/MMEB-train",
                "--output_dir", "logs/",
                "--logging_steps", "1",
                "--lr_scheduler_type", "linear",
                "--max_steps", "2000",
                "--warmup_steps", "200",
                "--save_steps", "1000",
                "--learning_rate", "2e-5",
                "--normalize", "True",
                "--temperature", "0.02",
                "--per_device_train_batch_size", "64",
                // "--max_restarts", "0",
                "--split_name", "original",
                "--image_resolution", "mid",
                "--max_len", "4096",
                "--save_safetensors", "False",
                "--remove_unused_columns", "False",
                "--report_to", "none",
                // "--grad_cache", "False",
                "--grad_cache", "True",
                "--gc_q_chunk_size", "8",
                "--gc_p_chunk_size", "8",
                "--lora",
                "--lora_r", "16",
                "--bf16", 
                "--pooling", "last"
                ],
        }
    ]
}`  Hello, this is my training config. The data I used for this test are MSCOCO_t2i and MSCOCO_i2t. Actually, when I use the full image dataset, the loss becomes zero from the beginning. I don’t understand why this is happening. ....

**Comment by PixarHero** [comment]
I use the v1 code, and the data loading code in v1 branch seems has some bugs, I just ignore the negtive samples.After investigating, I found that in gradcache, running the forward pass produces NaN representations. Do you have any advice on how to address this?

```
 def forward_no_grad(
            self,
            model: nn.Module,
            model_inputs,
    ) -> [Tensor, List[RandContext]]:
        """
        The first forward pass without gradient computation.
        :param model: Encoder model.
        :param model_inputs: Model input already broken into chunks.
        :return: A tuple of a) representations and b) recorded random states.
        """
        rnd_states = []
        model_reps = []
        with torch.no_grad():
            for x in model_inputs:
                rnd_states.append(RandContext(*self.get_input_tensors(x)))
                y = self.model_call(model, x)
                model_reps.append(self.get_reps(y))
        model_reps = torch.cat(model_reps, dim=0)
        return model_reps, rnd_states
```

**Comment by memray** [comment]
Hi @PixarHero ,

Thanks for following up on this. The v1 code has been archived and we don't plan to continue maintaining it. Would you please considering using the latest v2 code (data loading part has been refactored very much while modeling part is basically the same)?
Two specific suggestions:
1. remove --max_len since it may truncate the input unexpectedly.
2. set a breakpoint in debugging and see if '<|image_pad|> and EOS is in the input tensors.

Best,
Rui

**Comment by PixarHero** [comment]
Hi, could you share which Python environment (Python version, CUDA version, and the specific packages/versions) you used when it ran normally? This will help me reproduce the setup. Thanks!

> [@PixarHero](https://github.com/PixarHero) It might be a Python environment issue. I switched to a different environment and repackaged the code, and it ran normally.

**Comment by PixarHero** [comment]
Problem solved, it's because of the environment.
